sprinter/safe_buffer.go

```diff
@@ -0,0 +1,76 @@
+package sprinter
+
+import (
+	"io"
+	"sync"
+)
+
+type SafeBuffer struct {
+	mtx     sync.RWMutex
+	newData []chan int // notify readers that new data is available
+	done    bool       // set on close
+	data    []byte
+}
+
+func (b *SafeBuffer) Close() error {
+	b.mtx.Lock()
+	b.done = true
+	b.mtx.Unlock()
+
+	// inform the readers we're done
+	b.mtx.RLock()
+	for _, ch := range b.newData {
+		ch <- 0
+	}
+	b.mtx.RUnlock()
+	return nil
+}
+
+func (b *SafeBuffer) Write(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+	b.mtx.Lock()
+	b.data = append(b.data, p...)
+	b.mtx.Unlock()
+
+	// inform the readers how much new data we have
+	b.mtx.RLock()
+	for _, ch := range b.newData {
+		ch <- len(p)
```

Comment:
With this design for your signaling, you'll end up blocking everything if a single output writer (in `(*runnerImpl).StreamOutput`) ends up being slow or blocks.

There's also no way to remove a reader from `b.newData`, so any cancellation of `StreamOutput` will result in a channel in your list that's no longer being read from.